### ISSUE TITLE
KSQL should support 1:N Table-Table joins

### ISSUE THREAD

**Issue by rmoff** [confluentinc/ksql#1559]
Data model: 

*Customers* --N:1-- *Loyalty Club Status Levels*

A customer has one loyalty club status

A loyalty club status can have many customers

A common request would be 

* "Show me which customers are at 'platinum' level"
* "For each customer, show me their loyalty club status and associated details"

It's not possible to answer this with KSQL as it currently stands. Each is logically a KSQL TABLE (not stream), and so we use the table-table join. However, because each table must be partitioned on the join key for it to work, 1:N relationships don't work. 

Source data - note two customers at 'platinum' level: 

```
ksql> SELECT * FROM CUSTOMERS;
1530995903935 | 1 | 1 | Annemarie | Arent | aarent0@cpanel.net | Female | platinum | Organized web-enabled ability
1530995903935 | 2 | 2 | Merilyn | Doughartie | mdoughartie1@dedecms.com | Female | platinum | Optimized local definition
1530995903935 | 3 | 3 | Mirilla | Orwin | morwin2@ehow.com | Female | gold | Extended system-worthy superstructure
1530995903935 | 4 | 4 | Shirl | Barnfather | sbarnfather3@arizona.edu | Female | bronze | Vision-oriented global extranet
^CQuery terminated
ksql>
ksql>
ksql> SELECT * FROM CLUB;
1530996031951 | platinum | platinum | 100 | Top tier status.
1530996031951 | gold | gold | 50 | Mid tier status.
1530996031951 | bronze | bronze | 10 | Bottom tier status.
```

Which customers are at each level? 

```
ksql> SELECT X.CLUB_STATUS, C.FIRST_NAME + ' ' + C.LAST_NAME AS FULL_NAME FROM CLUB X LEFT JOIN CUSTOMERS C ON X.CLUB_STATUS=C.CLUB_STATUS;
platinum | null null
gold | null null
bronze | null null
```

Doesn't work, because `CUSTOMERS` isn't keyed on the join key `CLUB_STATUS` 

Reverse the join—For each customer, show details about their club status level: 

```
ksql> SELECT C.FIRST_NAME + ' ' + C.LAST_NAME AS FULL_NAME, C.CLUB_STATUS, X.CLUB_STATUS, X.MIN_ORDERS, X.COMMENTS FROM CUSTOMERS C LEFT JOIN CLUB X ON C.CLUB_STATUS = X.CLUB_STATUS;
Annemarie Arent | platinum | null | null | null
Merilyn Doughartie | platinum | null | null | null
Mirilla Orwin | gold | null | null | null
Shirl Barnfather | bronze | null | null | null
```

Doesn't work, because `CUSTOMERS` isn't keyed on the join key `CLUB_STATUS` 

Re-key the `CUSTOMERS` data, by `CLUB_STATUS`: 

```
CREATE STREAM CUSTOMERS_SRC (id BIGINT, first_name VARCHAR, last_name VARCHAR, email VARCHAR, gender VARCHAR, club_status VARCHAR, comments VARCHAR) WITH (KAFKA_TOPIC='customers', VALUE_FORMAT='JSON');
CREATE STREAM CUSTOMERS_BY_STATUS_STREAM WITH (VALUE_FORMAT='AVRO',KAFKA_TOPIC='CUST_BY_STAT2', PARTITIONS=1) AS SELECT * FROM CUSTOMERS_SRC PARTITION BY CLUB_STATUS;
CREATE TABLE CUSTOMERS_BY_STATUS WITH (KAFKA_TOPIC='CUST_BY_STAT2', VALUE_FORMAT='AVRO', KEY='CLUB_STATUS');
```

Note that there are now only three rows on the customer table, because it's showing the single value per key, and the key is no longer unique across all messages. 

```
ksql> SELECT * FROM CUSTOMERS_BY_STATUS;
1530995903935 | platinum | 2 | Merilyn | Doughartie | mdoughartie1@dedecms.com | Female | platinum | Optimized local definition
1530995903935 | gold | 3 | Mirilla | Orwin | morwin2@ehow.com | Female | gold | Extended system-worthy superstructure
1530995903935 | bronze | 4 | Shirl | Barnfather | sbarnfather3@arizona.edu | Female | bronze | Vision-oriented global extranet
```

So when we try and do the join, we don't see all the customers returned: 

```
ksql> SELECT X.CLUB_STATUS, C.FIRST_NAME + ' ' + C.LAST_NAME AS FULL_NAME FROM CLUB X LEFT JOIN CUSTOMERS_BY_STATUS C ON X.CLUB_STATUS=C.CLUB_STATUS;

platinum | Merilyn Doughartie
gold | Mirilla Orwin
bronze | Shirl Barnfather
platinum | Merilyn Doughartie
gold | Mirilla Orwin
bronze | Shirl Barnfather


ksql> SELECT C.FIRST_NAME + ' ' + C.LAST_NAME AS FULL_NAME, C.CLUB_STATUS, X.CLUB_STATUS, X.MIN_ORDERS, X.COMMENTS FROM CUSTOMERS_BY_STATUS C LEFT JOIN CLUB X ON C.CLUB_STATUS = X.CLUB_STATUS;
Merilyn Doughartie | platinum | platinum | 100 | Top tier status.
Mirilla Orwin | gold | gold | 50 | Mid tier status.
Shirl Barnfather | bronze | bronze | 10 | Bottom tier status.
Merilyn Doughartie | platinum | platinum | 100 | Top tier status.
Mirilla Orwin | gold | gold | 50 | Mid tier status.
Shirl Barnfather | bronze | bronze | 10 | Bottom tier status.
```

(not sure why we get double results either `¯\_(ツ)_/¯` )

**Comment by ZhiXingHeYiApple** [comment]
@rmoff That's the problem I'm facing. I think this usage should be very common, too. Why the PARTITION BY clause in KSQL will change the ROWKEY and specify the column as record's key? That make me confused. I think PARTITION BY should has no impact on record's key.
If now KSQL can't support 1:N denormalisation pattern, whether we can use kafka streams API to implement it? Some references will be highly appreciated! Thank you!

**Comment by JimGalasyn** [comment]
@rmoff, this is mentioned in KSQL [Table-Table join](https://docs.confluent.io/current/ksql/docs/tutorials/basics-docker.html#table-table-join) topic, but it seems important enough that it should also be mentioned somewhere in the KStreams [KTable-KTable Join](https://docs.confluent.io/current/streams/developer-guide/dsl-api.html#ktable-ktable-join) topic. What do you think?

**Comment by guozhangwang** [comment]
If the intention is to state that non-key KTable-KTable joins are not supported, I think it is fine. And there is an open KIP for supporting it on the roadmap: https://cwiki.apache.org/confluence/display/KAFKA/KIP-213+Support+non-key+joining+in+KTable

**Comment by JimGalasyn** [comment]
Tracking the KTable-KTable Join issue with internal ticket KSTREAMS-1659.

**Comment by rmoff** [comment]
Re-opening for community visibility

**Comment by fskuboadeo** [comment]
Hi everyone !!! I have faced the same problem to solve ... I think it would be great to have this functionality ... for us is one of the key decision points to use KSQL here in the project ...

**Comment by madnight** [comment]
Is there any workaround?

**Comment by mjsax** [comment]
Tracked via https://github.com/confluentinc/ksql/issues/4424. Closing this ticket as duplicate.
